### PR TITLE
Add a test for filtering distributions by content

### PR DESCRIPTION
**Pull request by lubosmj**
Required PR: https://github.com/pulp/pulpcore/pull/2992

[noissue]

### CODE REVIEW

**Comment by bmbouter** [comment]
@lubosmj please rebase. We'd like to move forward to merge. Thanks!

**Comment by lubosmj** [comment]
@bmbouter, I rebased the PR.

File: pulp_file/tests/functional/api/from_pulpcore/test_distributions.py [hunk 1]
```diff
@@ -162,3 +164,101 @@ def test_distribution_base_path(
...
+    data = FileFileDistribution(name="pub1", base_path="pub1", publication=publication.pulp_href)
+    distribution_pub1 = gen_object_with_cleanup(file_distro_api_client, data)
+
+    results = file_distro_api_client.list(with_content=content1.pulp_href).results
+    assert [distribution_pub1] == results
+
+    # test if the same publication attached to another distribution exposes the same content
+    data = FileFileDistribution(name="pub2", base_path="pub2", publication=publication.pulp_href)
+    distribution_pub2 = gen_object_with_cleanup(file_distro_api_client, data)
+
```
Comment (mdellweg) [inline pulp_file/tests/functional/api/from_pulpcore/test_distributions.py:206]: Can we attach a third distribution to the publication of the empty 0th repoversion?

File: pulp_file/tests/functional/api/from_pulpcore/test_distributions.py [hunk 2]
```diff
@@ -162,3 +164,101 @@ def test_distribution_base_path(
...
+    results = set(
+        d.pulp_href for d in file_distro_api_client.list(with_content=content2.pulp_href).results
+    )
+    assert {distribution_pub3.pulp_href, distribution_repopub.pulp_href} == results
+
+    # test if a repository without any attached publication does not expose any kind of content
+    # to a user even though the content is still present in the latest repository version
+    data = FileFileDistribution(name="repo", base_path="repo", repository=repo2.pulp_href)
+    _ = gen_object_with_cleanup(file_distro_api_client, data)
```
Comment (mdellweg) [inline pulp_file/tests/functional/api/from_pulpcore/test_distributions.py:255]:
```suggestion
    gen_object_with_cleanup(file_distro_api_client, data)
```

File: pulp_file/tests/functional/api/from_pulpcore/test_distributions.py [hunk 3]
```diff
@@ -162,3 +164,101 @@ def test_distribution_base_path(
...
+        content = file_content_api_client.list(repository_version_added=version_href).results[0]
+        return repo, content
+
+    repo1, content1 = generate_repo_with_content()
+
+    publish_data = FileFilePublication(repository=repo1.pulp_href)
+    publication = gen_object_with_cleanup(file_pub_api_client, publish_data)
+
+    # test if a publication attached to a distribution exposes the published content
+    data = FileFileDistribution(name="pub1", base_path="pub1", publication=publication.pulp_href)
```
Comment (mdellweg) [inline pulp_file/tests/functional/api/from_pulpcore/test_distributions.py:197]: You should be using something random for the name and base_path too make the test able to run in parallel.

File: pulp_file/tests/functional/api/from_pulpcore/test_distributions.py (hunk 2 above)
Comment (mdellweg) [reply pulp_file/tests/functional/api/from_pulpcore/test_distributions.py:255]: Why does this new distribution not pop up in the next list command?
It serves from repo2 latest version that should contain content2. No?

Comment (lubosmj) [reply pulp_file/tests/functional/api/from_pulpcore/test_distributions.py:255]: The plugin serves content only if there is a publication for a repository created. Content is not exposed when `SERVE_FROM_PUBLICATION` is set to `True` in the corresponding data model (https://github.com/pulp/pulp_file/search?q=SERVE_FROM_PUBLICATION) and the publication does not exist.

```
(pulp) [vagrant@pulp3-source-fedora35 ~]$ pulp file remote create --name foo --url https://fixtures.pulpproject.org/file/PULP_MANIFEST
{
  "pulp_href": "/pulp/api/v3/remotes/file/file/99478fd6-0986-4e29-8ba9-1b635ad58f4a/",
  "pulp_created": "2022-08-22T09:51:51.293969Z",
  "name": "foo",
  "url": "https://fixtures.pulpproject.org/file/PULP_MANIFEST",
  "ca_cert": null,
  "client_cert": null,
  "tls_validation": true,
  "proxy_url": null,
  "pulp_labels": {},
  "pulp_last_updated": "2022-08-22T09:51:51.294146Z",
  "download_concurrency": null,
  "max_retries": null,
  "policy": "immediate",
  "total_timeout": null,
  "connect_timeout": null,
  "sock_connect_timeout": null,
  "sock_read_timeout": null,
  "headers": null,
  "rate_limit": null
}

(pulp) [vagrant@pulp3-source-fedora35 ~]$ pulp file repository create --name foo --remote foo
{
  "pulp_href": "/pulp/api/v3/repositories/file/file/eb1c0b21-a996-4c6c-b1bb-ad57866b2a12/",
  "pulp_created": "2022-08-22T09:51:58.905572Z",
  "versions_href": "/pulp/api/v3/repositories/file/file/eb1c0b21-a996-4c6c-b1bb-ad57866b2a12/versions/",
  "pulp_labels": {},
  "latest_version_href": "/pulp/api/v3/repositories/file/file/eb1c0b21-a996-4c6c-b1bb-ad57866b2a12/versions/0/",
  "name": "foo",
  "description": null,
  "retain_repo_versions": null,
  "remote": "/pulp/api/v3/remotes/file/file/99478fd6-0986-4e29-8ba9-1b635ad58f4a/",
  "autopublish": false,
  "manifest": "PULP_MANIFEST"
}

(pulp) [vagrant@pulp3-source-fedora35 ~]$ pulp file repository sync --name foo
Started background task /pulp/api/v3/tasks/37f531ef-4bf2-4f41-9d18-80252d28b7c2/
..Done.
(pulp) [vagrant@pulp3-source-fedora35 ~]$ pulp file distribution create --name foo --base-path foo --repository foo
Started background task /pulp/api/v3/tasks/0e1bf233-b598-4ee1-80e3-c03540bfb54b/
.Done.
{
  "pulp_href": "/pulp/api/v3/distributions/file/file/2ef7195c-a820-45f9-867f-2dc82eb2af9f/",
  "pulp_created": "2022-08-22T09:52:22.406515Z",
  "base_path": "foo",
  "base_url": "https://pulp3-source-fedora35.localhost.example.com/pulp/content/foo/",
  "content_guard": null,
  "pulp_labels": {},
  "name": "foo",
  "repository": "/pulp/api/v3/repositories/file/file/eb1c0b21-a996-4c6c-b1bb-ad57866b2a12/",
  "publication": null
}

(pulp) [vagrant@pulp3-source-fedora35 ~]$ http https://pulp3-source-fedora35.localhost.example.com/pulp/content/foo/
HTTP/1.1 404 Not Found
Connection: keep-alive
Content-Length: 14
Content-Type: text/plain; charset=utf-8
Date: Mon, 22 Aug 2022 09:52:32 GMT
Server: nginx

404: Not Found
```

Comment (mdellweg) [reply pulp_file/tests/functional/api/from_pulpcore/test_distributions.py:255]: So what's the point in creating a Distribution with a Repository attached? Is this for auto-publish?

Comment (lubosmj) [reply pulp_file/tests/functional/api/from_pulpcore/test_distributions.py:255]: The point is just to test if the filtering procedure correctly lists only distributions that really expose the content. The content is not served by a distribution if there is no publication defined and `autopublish` is set to False.

**Comment by lubosmj** [comment]
@mdellweg, I have tried to address all the comments. I also added a test scenario where we can test if a publication created additionally for an old repository causes the distribution referencing that repository to expose the old content.

File: pulp_file/tests/functional/api/from_pulpcore/test_distributions.py (hunk 2 above)
Comment (mdellweg) [reply pulp_file/tests/functional/api/from_pulpcore/test_distributions.py]: Sorry i didn't mean the test this time. I was asking what good is a distribution with a repository. Or weather we should reject creating such a thing. (Unrelated to this feature)

File: pulp_file/tests/functional/api/from_pulpcore/test_distributions.py [hunk 4]
```diff
@@ -162,3 +164,119 @@ def test_distribution_base_path(
...
+    data = FileFileDistribution(
+        name="pub3", base_path="pub3", publication=new_publication.pulp_href
+    )
+    distribution_pub3 = gen_object_with_cleanup(file_distro_api_client, data)
+
+    results = set(
+        d.pulp_href for d in file_distro_api_client.list(with_content=content2.pulp_href).results
+    )
+    assert {distribution_pub3.pulp_href, distribution_repopub.pulp_href} == results
+
```
Comment (mdellweg) [inline pulp_file/tests/functional/api/from_pulpcore/test_distributions.py:254]: If we skipped this check in favor of the one 8 lines down. Would the test loose anything?

**Review by mdellweg — APPROVED** [review]
One picky question to maybe save a call in the testsuite left.

File: pulp_file/tests/functional/api/from_pulpcore/test_distributions.py (hunk 2 above)
Comment (lubosmj) [reply pulp_file/tests/functional/api/from_pulpcore/test_distributions.py]: Yes, we allow creating distributions with a repository, even though it does not make any sense.

File: pulp_file/tests/functional/api/from_pulpcore/test_distributions.py (hunk 4 above)
Comment (lubosmj) [reply pulp_file/tests/functional/api/from_pulpcore/test_distributions.py]: No. It was there just for readability purposes. I am removing it and adding a note for that.

**Pull request merged**